Design/purity-checker-proposal.md

```diff
@@ -0,0 +1,103 @@
+## !First Concept! for the Purity Checker
+
+We want a call of `checkValidity()` on an arbitrary `ExpEmbedding` node to return `true` if all `assert` statements present contain only pure expressions, and `false` as soon as one of them is impure. We split the logic into traversal and validation, currently only focussing on the validation of expressions appearing inside of `assert` nodes.
+
+Every `ExpEmbedding` implements a new `children` property and the functions `checkValidity` and `checkOwnValidity`.
+
+The default for `children` is an empty immutable list (`emptyList()`); when overridden it must still be an immutable list referencing the children’s subsequent embeddings.
+
+| **Embedding** | **children** |
+| --- | --- |
+| Assert | `exp` |
+| Block | `exps[]` |
+| Assign | `lhs`, `rhs` |
+| DirectResult | `subexpressions` |
+| Passthrough | `inner` |
+| If | `condition`, `thenBranch`, `elseBranch` |
+| While | `condition`, `body` |
+| MethodCall | `args` |
+
+---
+
+### checkValidity
+
+```kotlin
+fun checkValidity(): Boolean =
+    children().all { it.checkValidity() } && checkOwnValidity()
+
+```
+
+`checkValidity` is responsible only for traversing; the above is the default implementation every `ExpEmbedding` node inherits.
+
+---
+
+### checkOwnValidity
+
+```kotlin
+interface ExpEmbedding {
+    fun children(): List<ExpEmbedding> = emptyList()
+    fun checkOwnValidity(): Boolean = true
+}
+
+```
+
+The default local rule just returns `true`.
+
+---
+
+### Only validating `Assert`
+
+```kotlin
+data class Assert(val exp: ExpEmbedding) : ExpEmbedding {
+
+    override fun children(): List<ExpEmbedding> = listOf(exp)
+
+    override fun checkValidity(): Boolean =
+        checkOwnValidity()                       // stop traversal here
+
+    override fun checkOwnValidity(): Boolean =
+        PurityRules.isExpressionValid(exp)
+}
+
+```
+
+An `Assert` node ends the traversal and validates only itself.
+
+---
+
+### PurityRules
+
+`assert` can only contain embeddings evaluating to a boolean, so for the validation we can restrict ourselves to:
```

Comment:
Does it matter?  The subexpressions of these can still evaluate to something else.